### PR TITLE
Initial OpenMP Pragma support

### PR DESCRIPTION
**Pull request by Thirumalai-Shaktivel**
Towards: https://github.com/lfortran/lfortran/issues/3777

### CODE REVIEW

**Comment by Thirumalai-Shaktivel** [comment]
@certik, What do you think about this implementation? If there are any improvements or suggestions, do let me know; otherwise, I will add the required commit messages and tests.

**Comment by Thirumalai-Shaktivel** [comment]
This provides the multiline pragma's support, I need to figure how we need to handle singleline pragma's:
```fortran
     !$omp atomic
     n = n + 1
```

**Review by certik — APPROVED** [review]
I reviewed it, I think this is fine. The parser changes look good.

The only thing I don't know if PragmaBlock should be a block and contain what is in it, or rather it should be just a line, so one Pragma AST node for the beginning of the block and another for the end.

We will know which approach is better once we implement the AST->ASR pass.

So for now I think this is ok, and if we need to change it later, we can.

**Review by Pranavchiku — APPROVED** [review]
The node has sufficient information to construct `do concurrent` or any other `OMPNode` in ASR, we can later see how things unfold.

**Comment by Pranavchiku** [comment]
> a.f90

```fortran
subroutine simple(n, a, b)
integer :: i, n
real :: a(n), b(n)

!$omp parallel private(i) shared(n, a, b)
!$omp do
do i = 2, n
    b(i) = (a(i) + a(i-1)) / 2.0
end do
!$omp end do
!$omp end parallel

end subroutine
```

<details>

<summary> ast corresponding to pragma </summary>

```clojure
(PragmaBlock
    0
    "parallel"
    [(String
        "private(i)"
    )
    (String
        "shared(n,"
    )
    (String
        "a,"
    )
    (String
        "b)"
    )]
    [(PragmaBlock
        0
        "do"
        []
        [(DoLoop
            0
            ()
            0
            i
            2
            n
            ()
            [(Assignment
                0
                (FuncCallOrArray
                    b
                    []
                    [(()
                    i
                    ()
                    0)]
                    []
                    []
                    []
                )
                (/ (+ (FuncCallOrArray
                    a
                    []
                    [(()
                    i
                    ()
                    0)]
                    []
                    []
                    []
                ) (FuncCallOrArray
                    a
                    []
                    [(()
                    (- i 1)
                    ()
                    0)]
                    []
                    []
                    []
                )) (Real "2.0"))
                ()
            )]
            ()
            ()
        )]
        ()
    )]
    (TriviaNode
        []
        [(EndOfLine)
        (EndOfLine)]
    )
)
```

</details>

OpenMp have a few clauses listed:
* if(scalar-logical-expression)
* untied
* default(private|firstprivate|shared|none)
* private(list)
* firstprivate(list)
* shared(list)

I believe instead of defining clauses as `expr* clauses`, can't we tokenize each and have dedicated node for it like `do_loop_head` in ASR.

**Comment by certik** [comment]
@Thirumalai-Shaktivel what needs to be done here to finish it?

**Comment by Thirumalai-Shaktivel** [comment]
> OpenMp have a few clauses listed 
https://github.com/lfortran/lfortran/pull/3864#issuecomment-2054011743

With this PR changes, we store all the clauses as StringConstant. Currently I'm thinking how we handle this to create dedicated nodes.

Three options
- Create a tokenizer to recognise !$omp statement
- Recognise `default(private|firstprivate|shared|none)` as identifiers (this might leads us to more conflicts)
- Handles the strings in `semantics.h` file and create the required nodes.

**Comment by certik** [comment]
I think the tokenizer should recognize `!$omp` and emit `OMPPragma` token:
* Either it returns the rest of the line as a string, then in AST->ASR we parse the string, and if we use a recursive descent parser, we can easily and unambiguously parse it
* Or for the rest of the line it switches to a different mode and produces dedicated tokens. Then we parse it in the main parser, but it will likely produce conflicts for things like `default(default)`.
So I would go with the first option.

File: src/lfortran/semantics/ast_body_visitor.cpp [hunk 1]
```diff
@@ -3590,6 +3590,10 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
 
     }
 
+    void visit_Pragma(const AST::Pragma_t /*&x*/) {
+        // ignore for now
```
Comment (Thirumalai-Shaktivel) [inline src/lfortran/semantics/ast_body_visitor.cpp:3594]: We skip the OMPPragma in the AST->ASR for now.

**Comment by Thirumalai-Shaktivel** [comment]
Go ahead review the changes and saved AST. If everything looks good, do merge it.

**Review by certik — APPROVED** [review]
I think this looks great, thanks!

File: tests/reference/ast-pragma1-6b56efa.stdout [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+            .false.
+            "parallel"
+            [(String
+                "private(partial_sum)"
+            )
+            (String
+                "shared(total_sum)"
+            )]
+            ()
+        )
```
Comment (Pranavchiku) [inline tests/reference/ast-pragma1-6b56efa.stdout:188]: this is good, easy to access variables

File: src/lfortran/semantics/ast_body_visitor.cpp (hunk 1 above)
Comment (Pranavchiku) [reply src/lfortran/semantics/ast_body_visitor.cpp:3594]: Yes, we need to create a DoConcurrent loop here, I know what to do, AST looks good, thanks for doing this!

**Pull request merged**